use replicas when initializing pg minResources

sceneryback opened the pull request:
When creating pod using replicaSet, the minMember of podgroup created is always 1, this is not reasonable. Consider this case:
1. we created a deployment with 2 replicas, totally requesting 16 gpus 
2. the podgroup created has 1 minMember and 8 gpus
3. the capability of queue has 8 gpus and the scheduler configmap disables  AllocatableFn of proportion plugin
4. the podgroup will be enqueued and running since minResources(8 gpus) is not bigger than queue capability
5. finally the queue allocated 16 gpus, more than its deserved

As can be seen from this picture:   
![image](https://github.com/user-attachments/assets/a3e3041d-44d2-4f3d-b948-005ab0e227ca)

In the case that pod has replicas from its owners, the podgroup should consider relicas, just as training-operator does, which creates pod group based on resources from all replicas

sceneryback commented:
/assign @hzxuzhonghu

hwdef commented:
Same situation as mentioned in this issue:
https://github.com/volcano-sh/volcano/issues/3970

Monokaix commented:
proportion plugin controls the queue related capabilities, if you disable it, the capacity check will not take effect.

Monokaix commented:
Just as @hwdef said, please set an annotation to control the minmember to set, and also other workload type like statefulset should also has this capability: )

sceneryback commented:
hi @hwdef @Monokaix thanks for all your suggestions, I have updated the feature using new annotation, so it won't break current behaviors. Please have a look.     
apis change: https://github.com/volcano-sh/apis/pull/154
I will update go.mod to delete local references of apis package after https://github.com/volcano-sh/apis/pull/154 is merged

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +181,37 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+
+	for _, reference := range pod.OwnerReferences {
+		if reference.Kind != "" && reference.Name != "" {
+			tmp := pg.getAnnotationsFromUpperRes(reference.Kind, reference.Name, pod.Namespace)
```
Usually we write annotation on the deployment to specify minMember, but here the ownerReferences of the Pod can only be traced back to the replicaset.

sceneryback replied:
From k8s source code https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L235, the ReplicaSet will inherit annotations from Deployment, so I think it's enough to get annotations from RS
![image](https://github.com/user-attachments/assets/4f7cf2b5-5858-4aa2-b906-5cdbd3b88762)

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +181,37 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+	for _, reference := range pod.OwnerReferences {
+		if reference.Kind != "" && reference.Name != "" {
+			tmp := pg.getAnnotationsFromUpperRes(reference.Kind, reference.Name, pod.Namespace)
+			if err := mergo.Merge(&annotations, &tmp); err != nil {
```
Do we have scenarios that Pod has multiple OwnerReferences? If there are trully multiple annotations, we won't specify twice the minMember annotations, do we need to use the merge.Merge?

sceneryback replied:
Normally there won't be multiple OwnerReferences with this annotation, updated. Thanks

hwdef replied:
Can we not import the Mergo package? Feeling little necessary

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +181,37 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
 }
 
+func (pg *pgcontroller) getMinMemberFromUpperRes(pod *v1.Pod) *int32 {
+	defaultMinMember := int32(1)
```
This variable name is very strange. You can set 1 as a constant (as defaultMinMember), and then the variable here is initialized to defaultMinMember. `minMember` here is ok

sceneryback replied:
Updated

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -193,7 +228,7 @@ func (pg *pgcontroller) inheritUpperAnnotations(pod *v1.Pod, obj *scheduling.Pod
 }
 
-func (pg *pgcontroller) createNormalPodPGIfNotExist(pod *v1.Pod) error {
+func (pg *pgcontroller) createNormalPodPGIfNotExist(pod *v1.Pod, minAvailable *int32) error {
```
Why are the variable names here not unified as minMember?

sceneryback replied:
Updated

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -203,6 +238,11 @@ func (pg *pgcontroller) createNormalPodPGIfNotExist(pod *v1.Pod) error {
 			return err
 		}
 
+		minMember := int32(1)
```
There is no need to initialize again here, you already have the initialization value in `getMinMemberFromUpperRes`.

sceneryback replied:
updated

JesseStutler commented on pkg/controllers/job/job_controller_util.go:
```diff
@@ -327,10 +327,10 @@ func (p TasksPriority) CalcFirstCountResources(count int32) v1.ResourceList {
 	for _, task := range p {
 		if count <= task.Replicas {
-			minReq = quotav1.Add(minReq, calTaskRequests(&v1.Pod{Spec: task.Template.Spec}, count))
+			minReq = quotav1.Add(minReq, *util.CalTaskRequests(&v1.Pod{Spec: task.Template.Spec}, count))
```
The return value of the function here is a pointer, and then you use `*` directly to get the value pointed to by the pointer, the way it's written looks a bit strange and doesn't quite fit in with go's clean code. The return value of the func `CalTaskRequests` does not need to be refactored, keeping the original is better. In fact, when the API was designed, minResources should not be a pointer, ResourceList is a map.

sceneryback replied:
simply copied from an existing util function:   
![image](https://github.com/user-attachments/assets/65d14edc-f2e1-448f-8f21-2c3c46456598)
I will update `GetPodQuotaUsage` too    
Also notice that the minResources field of PodGroup is designed as pointer originally, so currently won't break this

sceneryback commented:
hi @JesseStutler Thanks for your reviewing! Have updated, please have a look

Monokaix commented:
API related pr is merged, you can move on: )

sceneryback commented:
Updated go.mod, please have a look @Monokaix @hwdef @JesseStutler

hwdef reviewed (commented):
/lgtm

hwdef commented:
/ok-to-test

Monokaix commented on pkg/controllers/util/util.go:
```diff
@@ -4,17 +4,28 @@ import (
 )
 
-func GetPodQuotaUsage(pod *v1.Pod) *v1.ResourceList {
+func GetPodQuotaUsage(pod *v1.Pod) v1.ResourceList {
```
Why changed to non-poiner?

sceneryback replied:
Just follow this comment from @JesseStutler for better style https://github.com/volcano-sh/volcano/pull/4000#discussion_r1948988135

Monokaix commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -172,7 +172,8 @@ func (pg *pgcontroller) processNextReq() bool {
 	// normal pod use volcano
 	klog.V(4).Infof("Try to create podgroup for pod %s/%s", pod.Namespace, pod.Name)
-	if err := pg.createNormalPodPGIfNotExist(pod); err != nil {
+	minMember := pg.getMinMemberFromUpperRes(pod)
```
This logic should put in `createNormalPodPGIfNotExist`, because when the podgroup is already existed, there is no need to calculate the min member, it will request APIServer which is a heavy operation.
And there is also a qustion mentioned here https://github.com/volcano-sh/volcano/issues/3970#issuecomment-2652862212

sceneryback replied:
very good point 👍 so it's better to keep rs/sts etc informers in pgcontroller and read cache from these informers, right?

Monokaix replied:
Yeah, DaemonSet and Job should also be considered.

Monokaix replied:
Also this should also be considered https://github.com/volcano-sh/volcano/issues/3970#issuecomment-2638750095

sceneryback replied:
> Also this should also be considered [#3970 (comment)](https://github.com/volcano-sh/volcano/issues/3970#issuecomment-2638750095)

I think it's not necessary to check pod annotation? The annotation is designed for workload, it will be strange to see this in pod

sceneryback commented:
Refactored code using informers, please have a look again @hwdef @Monokaix @JesseStutler

hwdef commented:
Hi, please make the CI happy

sceneryback commented:
hi @hwdef fixed the failed e2e tests, please have a look again

hwdef reviewed (commented):
/lgtm
Thanks!

hwdef commented:
PTAL
@Monokaix @JesseStutler @lowang-bh

Monokaix commented on go.mod:
```diff
@@ -46,7 +46,7 @@ require (
 	sigs.k8s.io/yaml v1.4.0
 	stathat.com/c/consistent v1.0.0
-	volcano.sh/apis v1.10.0-alpha.0.0.20241210014034-bf27f4e986d0
+	volcano.sh/apis v1.11.1-0.20250211082520-7f8222e881d9
```
Why 1.11.1?

Monokaix commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -99,15 +118,29 @@ func (pg *pgcontroller) Initialize(opt *framework.ControllerOption) error {
 		AddFunc: pg.addPod,
 	})
 
+	pg.rsInformer = pg.informerFactory.Apps().V1().ReplicaSets()
```
These should be under statement `if utilfeature.DefaultFeatureGate.Enabled(features.WorkLoadSupport) {`.

sceneryback replied:
rsLister is also used in 'inheritUpperAnnotations()' when WorkLoadSupport is not enabled, so only add these codes under the statement:   
```
pg.rsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
    AddFunc:    pg.addReplicaSet,
    UpdateFunc: pg.updateReplicaSet,
})
```

Monokaix replied:
WorkLoadSupport feature means whether Volcano supports deployment/rs/ds/sts,etc., so the informers should be initiated when featuregate is not true.

JesseStutler replied:
Do you mean to add another gate called `NonVcjobMinMemberSupport` to control whether to use informers for deployment/rs/sts/job? But `WorkLoadSupport` is used to control whether to support these resources

sceneryback replied:
> Do you mean to add another gate called `NonVcjobMinMemberSupport` to control whether to use informers for deployment/rs/sts/job? But `WorkLoadSupport` is used to control whether to support these resources

No, just add a param arg following the comments of @Monokaix to enable this feature

JesseStutler replied:
So why not use WorkloadSupport uniformly?

Monokaix commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -99,15 +118,29 @@ func (pg *pgcontroller) Initialize(opt *framework.ControllerOption) error {
 		AddFunc: pg.addPod,
 	})
 
+	pg.rsInformer = pg.informerFactory.Apps().V1().ReplicaSets()
```
I think we should add a param feature switch to users whether enable non-vcjob supports minMember, becasue it's a little heavy.

sceneryback replied:
hi @Monokaix you mean a feature gate like features.NonVcjobMinMemberSupport which only applies logic in this PR when enabled?

Monokaix replied:
a param flag is ok.

hwdef replied:
@sceneryback 
The nonvcjobminmembersupport flag is not very useful in the current code implementation.
I think @Monokaix means that if `nonvcjobminmembersupport` is true, then enable these informers, otherwise do not enable these informers because he thinks these informers are a bit heavy


@Monokaix Is it correct that I understand?

Monokaix replied:
Yaeh that's exactly what I mean.

Monokaix replied:
I forget that there is already a flag `inheritOwnerAnnotations` to get annotations from upper resource, so for simplicity, just re-use this flag and no need to add a new flag `nonvcjobminmembersupport` and no informer is neeed.

Monokaix replied:
> I forget that there is already a flag `inheritOwnerAnnotations` to get annotations from upper resource, so for simplicity, just re-use this flag and no need to add a new flag `nonvcjobminmembersupport` and no informer is neeed.

Please modify based on this, it will be clearer and simpler.

sceneryback replied:
you mean 1. drop this new flag 'NonVcjobMinMemberSupport'  2. drop the informers even if WorkloadSupport is true or init informers when WorkloadSupport is true?     I think WorkloadSupport is more proper than inheritOwnerAnnotations to enable this feature  @Monokaix @hwdef @JesseStutler 

> > I forget that there is already a flag `inheritOwnerAnnotations` to get annotations from upper resource, so for simplicity, just re-use this flag and no need to add a new flag `nonvcjobminmembersupport` and no informer is neeed.
> 
> Please modify based on this, it will be clearer and simpler.

Monokaix replied:
Yes use the `inheritOwnerAnnotations` is ok, and the feature we claim to the public is still non-vcjob support for setting minmember, while the detailed setting is to turn on this switch, the feature name and flag name do not need to be exactly the same. `WorkloadSupport` is used to control whether list/watch, we don't need informer any more on this feature.

Monokaix commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,33 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+					klog.Infof("Failed to convert minMemberAnnotation of Pod owners <%s/%s> into number, skipping",
+						pod.Namespace, pod.Name)
+					return minMember
+				}
```
< 0 is ok?

sceneryback replied:
should be positive, will fix

Monokaix commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,33 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+			if minMemberAnno, ok := annotations[scheduling.VolcanoGroupMinMemberAnnotationKey]; ok {
+				minMemberFromAnno, err := strconv.Atoi(minMemberAnno)
+				if err != nil {
+					klog.Infof("Failed to convert minMemberAnnotation of Pod owners <%s/%s> into number, skipping",
```
This klog info is not so clear, we should mention that remain minmember =1, or users will don't  know the value finally set.

Monokaix commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,33 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+					return minMember
+				}
+				if minMemberFromAnno < math.MinInt32 || minMemberFromAnno > math.MaxInt32 {
+					klog.Infof("minMember %d exceeds bounds of int32, skipping", minMember)
```
same above.

JesseStutler commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -99,15 +118,29 @@ func (pg *pgcontroller) Initialize(opt *framework.ControllerOption) error {
 
+	pg.rsInformer = pg.informerFactory.Apps().V1().ReplicaSets()
+	pg.rsLister = pg.rsInformer.Lister()
+	pg.rsSynced = pg.rsInformer.Informer().HasSynced
```
There is no need to add these rsSynced funcs, these codes:
```
	for informerType, ok := range pg.informerFactory.WaitForCacheSync(stopCh) {
		if !ok {
			klog.Errorf("caches failed to sync: %v", informerType)
		}
	}
	for informerType, ok := range pg.vcInformerFactory.WaitForCacheSync(stopCh) {
		if !ok {
			klog.Errorf("caches failed to sync: %v", informerType)
			return
		}
	}
```
has already done this

JesseStutler commented on installer/helm/chart/volcano/templates/controllers.yaml:
```diff
@@ -74,13 +74,13 @@ rules:
   - apiGroups: ["apps"]
     resources: ["daemonsets", "statefulsets"]
-    verbs: ["get"]
+    verbs: ["get", "list"]
```
If we change all these APIs to list/watch them, I think the verbs here should be ["list","watch"]?

JesseStutler commented:
Better to post a screenshot to verify your codes work in actual environment thanks

Monokaix commented:
Hi, are you still working on this? It's a nice feature and hope you can resolve the comments: )

hwdef commented:
Please rebase the master branch.
This PR is important.

sceneryback commented:
sorry for recent inactivity, I will continue working on this feature

hwdef commented:
Never mind. Looking forward to your new code. :)

sceneryback commented:
> Better to post a screenshot to verify your codes work in actual environment thanks

deployment with min-member as 2:   
![image](https://github.com/user-attachments/assets/5ba01ae5-75f8-4cde-a91c-6cd71d55c250)
deployment with min-member as 3:
![image](https://github.com/user-attachments/assets/b0330db1-f8b1-43a9-927d-d1af3d0d7344)
deployment with min-member as 0:
![image](https://github.com/user-attachments/assets/ad9c2fab-8c5f-421d-a5d2-058827188e25)
daemonset with min-member as 2:
![image](https://github.com/user-attachments/assets/2fe528d4-02f4-4ede-97ae-c1724ebf523d)
job with min-member as 2:
![image](https://github.com/user-attachments/assets/7bd4697e-53c3-4ea8-bb3b-9f7d0589460f)
statefulset with min-member as 2:
![image](https://github.com/user-attachments/assets/8ed692c8-f5b9-49a9-ac88-7ac88239f3af)

sceneryback commented:
Hi @hwdef @Monokaix @JesseStutler Updated code according to previous comments, mainly including:  
1. add a param flag 'non-vcjob-minmember-support' to enable this feature which is true by default
2. add more validation for this anno, especially deployment, statefulset and job
3. test this feature in local k8s cluster and provide screenshots for each kind

Please have a look again, thanks!

hwdef commented:
Excellent! I'll check this later.

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+			if minMemberAnno, ok := annotations[scheduling.VolcanoGroupMinMemberAnnotationKey]; ok {
+				minMemberFromAnno, err := strconv.Atoi(minMemberAnno)
+				if err != nil {
+					klog.Infof("Failed to convert minMemberAnnotation of Pod owners <%s/%s> into number, minMember remains as 1",
```
klog.Errof?

sceneryback replied:
fixed, thanks

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+					return minMember
+				}
+				if minMemberFromAnno <= 0 {
+					klog.Infof("minMemberAnnotation %d is not positive, minMember remains as 1", minMemberFromAnno)
```
Same as above, need to use klog.Errof

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+			klog.Errorf("Failed to get upper %s for Pod <%s/%s>: %v", kind, namespace, name, err)
+			return false
+		}
+		return *rs.Spec.Replicas >= int32(minMemberFromAnno)
```
Why do we need this verification? The verification here in the scheduler is not noticeable to the users, this is something that should be handled by the webhook manager

JesseStutler replied:
Need to discuss should we verify this if replicas of workload is larger than minMemberFromAnno? @Monokaix @hwdef

sceneryback replied:
If minMemberFromAnno is larger than replicas, it will fail the validFunc in gang plugin: 
```
		vtn := job.ValidTaskNum()
		if vtn < job.MinAvailable {
			return &api.ValidateResult{
				Pass:   false,
				Reason: v1beta1.NotEnoughPodsReason,
				Message: fmt.Sprintf("Not enough valid tasks for gang-scheduling, valid: %d, min: %d",
					vtn, job.MinAvailable),
			}
		}
```
e.g. replicas is 2 and minMemberFromAnno is 3, it will show 'Not enough valid tasks for gang-scheduling, valid: 2, min: 3' and the job will never be scheduled successfully

hwdef replied:
> Why do we need this verification? The verification here in the scheduler is not noticeable to the users, this is something that should be handled by the webhook manager

I recommend doing verification in the controller as well, because the current trend is to remove webhooks to enhance performance https://github.com/volcano-sh/volcano/issues/4127

JesseStutler replied:
> If minMemberFromAnno is larger than replicas, it will fail the validFunc in gang plugin:
> 
> ```
> 		vtn := job.ValidTaskNum()
> 		if vtn < job.MinAvailable {
> 			return &api.ValidateResult{
> 				Pass:   false,
> 				Reason: v1beta1.NotEnoughPodsReason,
> 				Message: fmt.Sprintf("Not enough valid tasks for gang-scheduling, valid: %d, min: %d",
> 					vtn, job.MinAvailable),
> 			}
> 		}
> ```
> 
> e.g. replicas is 2 and minMemberFromAnno is 3, it will show 'Not enough valid tasks for gang-scheduling, valid: 2, min: 3' and the job will never be scheduled successfully

If the user misconfigures it, they should be able to notice it. If it defaults to 1 when misconfigured, the user won't even notice

Monokaix replied:
Just keep the minMember value users specified, if minMember > replica, job will be in Pending and users can be aware of that.

JesseStutler commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+			return false
+		}
+		return *rs.Spec.Replicas >= int32(minMemberFromAnno)
+	case "StatefulSet", "Job":
```
Why do Statefulset and job not need it?

sceneryback replied:
Because I think it's meaningless for Statefulset (pod created 1 by 1) and job (1 pod), you cannot set a minMemberFromAnno larger than 1, or the ss and job will fail the validFunc in gang plugin for the same reason as before. Use the default 1 is enough for ss and job.

JesseStutler replied:
The pod of a job may not require gang scheduling, but the pod of an sts might still need it

sceneryback replied:
> sts

It's not easy to check the existence of gang plugin in controller code, so keep the current behavior is OK? i.e. setting minMember to 1 for job and sts @JesseStutler

Monokaix replied:
Sts can also create pods concurrently, and job can has multi replicas. We should support this feature for sts and job, and user can choose whether to config it according to their own use case.

Monokaix commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -124,6 +126,7 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet, knownControllers []string) {
 	fs.BoolVar(&s.EnableMetrics, "enable-metrics", false, "Enable the metrics function; it is false by default")
 	fs.StringVar(&s.ListenAddress, "listen-address", defaultListenAddress, "The address to listen on for HTTP requests.")
 	fs.BoolVar(&s.InheritOwnerAnnotations, "inherit-owner-annotations", true, "Enable inherit owner annotations for pods when create podgroup; it is enabled by default")
+	fs.BoolVar(&s.NonVcjobMinMemberSupport, "non-vcjob-min-member-support", true, "NonVcjobMinMemberSupport can initialize podgroup minMember and minResources for non-vcjobs(deployment/statefulset/daemonset/job).")
```
default value = false is better.

Monokaix commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+					klog.Errorf("minMemberAnnotation %d is not positive, minMember remains as 1", minMemberFromAnno)
+					return minMember
+				}
+				if minMemberFromAnno < math.MinInt32 || minMemberFromAnno > math.MaxInt32 {
```
You already used `strconv.Atoi` to parse a int number, so it's unnecessary to check the integer scope here.

sceneryback replied:
This checking is meant to avoid overflow because finally the value is assigned to minMember (int32)      
```
minMember = int32(minMemberFromAnno)
```

Monokaix replied:
If it overflow, the `strconv.Atoi` will return an err, that's to say, the validation has already been done.

sceneryback replied:
@Monokaix  If will fail the CodeQL: 
![image](https://github.com/user-attachments/assets/ceb3e7c4-e4bc-4890-b932-b358d5e18eee)

Monokaix replied:
We can use `minMember,err := strconv.ParseInt(str,10,32)` instead of `strconv.Atoi`.

Monokaix commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -57,14 +57,9 @@ type pgcontroller struct {
 
 	// A store of pods
 	podLister corelisters.PodLister
-	podSynced func() bool
```
We should keep this.

Monokaix commented:
> 1. reverted back to the none-informer's version
> 2. left the logic in isMinMemberAnnotationValidForKind unchanged, it may still need some discussion.
> 
> Please have a look again, thanks! @hwdef @Monokaix @JesseStutler

see https://github.com/volcano-sh/volcano/pull/4000/files#r2043332608

Monokaix commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -177,6 +179,58 @@ func (pg *pgcontroller) getAnnotationsFromUpperRes(kind string, name string, nam
+						pod.Namespace, pod.Name, err)
+					return minMember
+				}
+				if minMemberFromAnno <= 0 {
```
=0 is also valid.

Monokaix commented:
From my side, just fix the left comments: ), other parts lgtm.

sceneryback commented:
> > 1. reverted back to the none-informer's version
> > 2. left the logic in isMinMemberAnnotationValidForKind unchanged, it may still need some discussion.
> > 
> > Please have a look again, thanks! @hwdef @Monokaix @JesseStutler
> 
> see https://github.com/volcano-sh/volcano/pull/4000/files#r2043332608

Thanks for your suggestions! fixed

sceneryback commented:
@hwdef @Monokaix @JesseStutler updated code according to latest comments, please have a look again

Monokaix commented:
Well done! Nearly finished.

hwdef reviewed (commented):
/lgtm

sceneryback commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -203,6 +231,11 @@ func (pg *pgcontroller) createNormalPodPGIfNotExist(pod *v1.Pod) error {
 
+		var minMember = int32(1)
+		if pg.inheritOwnerAnnotations {
+			minMember = pg.getMinMemberFromUpperRes(pod)
```
fixed, thanks

sceneryback commented:
changed signatures of some functions for variable reuse, please have a look again @Monokaix @hwdef @JesseStutler

hwdef commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -191,52 +193,78 @@ func (pg *pgcontroller) updatePodAnnotations(pod *v1.Pod, pgName string) error {
+				tmp = job.Annotations
+			}
+
+			if err := mergo.Merge(&annotations, &tmp); err != nil {
```
Can we not import the Mergo package? Feeling little necessary

sceneryback replied:
or simply write our own map merging func, e.g: 
```
func mergeMap(src, dest map[string]string) {
    for k, v := range dest {
        if _, ok := src[k]; !ok {
            src[k] = v
        }
    }
}
```
or just use an inline one:   
```
    for k, v := range tmp {
        if _, ok := annotations[k]; !ok {
            annotations[k] = v
        }
    }
```
is this OK? @hwdef

hwdef replied:
It's good:)

sceneryback replied:
fixed, please have a look @hwdef

hwdef commented:
/lgtm
Thanks!

hwdef commented:
@Monokaix @JesseStutler @lowang-bh 
PTAL. Let's strike while the iron is hot.

hwdef commented:
/lgtm

Monokaix commented:
Overall ok to me, please keep one commit.

Monokaix commented:
/approve

hwdef commented:
/lgtm

Greate work!
